# **Pandas Introduction**
# What is Pandas?

Pandas is a Python library used for working with data sets.

It has functions for analyzing, cleaning, exploring, and manipulating data.

The name "Pandas" has a reference to both "Panel Data", and "Python Data Analysis" and was created by Wes McKinney in 2008.

\

# Why Use Pandas?
Pandas allows us to analyze big data and make conclusions based on statistical theories.

Pandas can clean messy data sets, and make them readable and relevant.

Relevant data is very important in data science.

\

# What Can Pandas Do?
Pandas gives you answers about the data. Like:

Is there a correlation between two or more columns?
What is average value?
Max value?
Min value?
Pandas are also able to delete rows that are not relevant, or contains wrong values, like empty or NULL values. This is called cleaning the data.

\

**Where is the Pandas Codebase?**

The source code for Pandas is located at this github repository https://github.com/pandas-dev/pandas

**Import Pandas**

Once Pandas is installed, import it in your applications by adding the import keyword:

In [ ]:
import pandas

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pandas.DataFrame(mydataset)

print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


**Pandas as pd**

Pandas is usually imported under the pd alias.

alias: In Python alias are an alternate name for referring to the same thing.

Create an alias with the as keyword while importing:

In [ ]:
import pandas as pd

# Now the Pandas package can be referred to as pd instead of pandas.


import pandas as pd

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pd.DataFrame(mydataset)

print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


**Pandas Series**

**What is a Series?**

A Pandas Series is like a column in a table.

It is a one-dimensional array holding data of any type.

In [ ]:
# Create a simple Pandas Series from a list:

import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a)

print(myvar)

0    1
1    7
2    2
dtype: int64


**Labels**

If nothing else is specified, the values are labeled with their index number. First value has index 0, second value has index 1 etc.

This label can be used to access a specified value.

In [ ]:
# Return the first value of the Series:

print(myvar[0])

1


**Create Labels**

With the index argument, you can name your own labels.

In [ ]:
# Create your own labels:

import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a, index = ["x", "y", "z"])

print(myvar)

x    1
y    7
z    2
dtype: int64


When you have created labels, you can access an item by referring to the label.


In [ ]:
# Return the value of "y":

print(myvar["y"])

7


In [4]:
# Create your own labels for this 2d array:

import pandas as pd

arr = [[1, 2, 3], [4, 5, 6]]

# Code Here
row = ['row1', 'row2']
col = ['col1', 'col2', 'col3']

# Create DataFrame
df = pd.DataFrame(arr, index=row, columns=col)

print(df)

      col1  col2  col3
row1     1     2     3
row2     4     5     6


In [5]:
# Return the value 5:

# Code Here
value = df.loc['row2', 'col2']
print(value)


5


**Key/Value Objects as Series**

You can also use a key/value object, like a dictionary, when creating a Series.

In [ ]:
# Create a simple Pandas Series from a dictionary:

import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories)

print(myvar)


day1    420
day2    380
day3    390
dtype: int64


Note: The keys of the dictionary become the labels.

To select only some of the items in the dictionary, use the index argument and specify only the items you want to include in the Series.

In [ ]:
# Create a Series using only data from "day1" and "day2":

import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories, index = ["day1", "day2"])

print(myvar)

day1    420
day2    380
dtype: int64


In [6]:
# Create a Series using only data from "b":

import pandas as pd

calories = {"a": 12, "b": 31, "c": 24}

# Code Here
myvar = pd.Series(calories["b"], index=["b"])

print(myvar)


b    31
dtype: int64


**DataFrames**

Data sets in Pandas are usually multi-dimensional tables, called DataFrames.

Series is like a column, a DataFrame is the whole table.

In [ ]:
# Create a DataFrame from two Series:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

myvar = pd.DataFrame(data)

print(myvar)

   calories  duration
0       420        50
1       380        40
2       390        45


**What is a DataFrame?**

A Pandas DataFrame is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.

In [8]:
# Create a simple Pandas DataFrame:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

print(df)

   calories  duration
0       420        50
1       380        40
2       390        45


**Locate Row**

As you can see from the result above, the DataFrame is like a table with rows and columns.

Pandas use the loc attribute to return one or more specified row(s)

In [9]:
# Return row 0:

#refer to the row index:
print(df.loc[0])

calories    420
duration     50
Name: 0, dtype: int64


In [10]:
# Return row 0 and 1:

#use a list of indexes:
print(df.loc[[0, 1]])

   calories  duration
0       420        50
1       380        40


In [11]:
# Return row 1 and 2:

#use a list of indexes:

# Code Here
print(df.loc[[1, 2]])

   calories  duration
1       380        40
2       390        45


**Named Indexes**

With the index argument, you can name your own indexes.

In [18]:
# Add a list of names to give each row a name:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

df = pd.DataFrame(data, index = ["day1", "day2", "day3"])

print(df)

      calories  duration
day1       420        50
day2       380        40
day3       390        45


In [21]:
# Add a list of names to give each row a name:

import pandas as pd

data1 = {
  "a": [3, 6, 9, 12],
  "b": [2, 4, 6, 8],
  "c": [10, 20, 30, 50]
}

# Code Here

df1 = pd.DataFrame(data1, index = ["row1", "row2", "row3", "row4"])

print(df1)

       a  b   c
row1   3  2  10
row2   6  4  20
row3   9  6  30
row4  12  8  50


**Locate Named Indexes**

Use the named index in the loc attribute to return the specified row(s).

In [ ]:
# Return "day2":

# refer to the named index:
print(df.loc["day2"])

calories    380
duration     40
Name: day2, dtype: int64


In [23]:
# Return a row from df1:

# refer to the named index:

# Code Here
print(df1.loc["row1"])

a     3
b     2
c    10
Name: row1, dtype: int64


**Pandas Read CSV**

**Read CSV Files**

A simple way to store big data sets is to use CSV files (comma separated files).

CSV files contains plain text and is a well know format that can be read by everyone including Pandas.

In our examples we will be using a CSV file called 'data.csv'.

In [25]:
# Load the CSV into a DataFrame:

import pandas as pd

df = pd.read_csv('data.csv')

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

In [26]:
# Load the "forestfires.csv" into a DataFrame:

# Code Here

import pandas as pd

df1 = pd.read_csv('forestfires.csv')

print(df1.to_string())

     X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain     area
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0     0.00
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0     0.00
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0     0.00
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2     0.00
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0     0.00
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0     0.00
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0     0.00
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0     0.00
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0     0.00
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0     0.00
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0     0.00
11   7  5   sep  sat  92.8  

use to_string() to print the entire DataFrame.

If you have a large DataFrame with many rows, Pandas will only return the first 5 rows, and the last 5 rows:


In [ ]:
# Print the data DataFrame without the to_string() method:

import pandas as pd

print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


In [27]:
# Print the "forestfires" DataFrame without the to_string() method:

# Code Here
import pandas as pd

print(df1)

     X  Y month  day  FFMC    DMC     DC   ISI  temp    RH  wind  rain   area
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2  51.0   6.7   0.0   0.00
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0  33.0   0.9   0.0   0.00
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6  33.0   1.3   0.0   0.00
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3  97.0   4.0   0.2   0.00
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4  99.0   1.8   0.0   0.00
..  .. ..   ...  ...   ...    ...    ...   ...   ...   ...   ...   ...    ...
512  4  3   aug  sun  81.6   56.7  665.6   1.9  27.8  32.0   2.7   0.0   6.44
513  2  4   aug  sun  81.6   56.7  665.6   1.9  21.9  71.0   5.8   0.0  54.29
514  7  4   aug  sun  81.6   56.7  665.6   1.9  21.2  70.0   6.7   0.0  11.16
515  1  4   aug  sat  94.4  146.0  614.7  11.3  25.6  42.0   4.0   0.0   0.00
516  6  3   nov  tue  79.5    3.0  106.7   1.1  11.8  31.0   4.5   0.0   0.00

[517 rows x 13 columns]


max_rows
The number of rows returned is defined in Pandas option settings.

You can check your system's maximum rows with the pd.options.display.max_rows statement.

In [ ]:
# Check the number of maximum returned rows:

import pandas as pd

print(pd.options.display.max_rows)

60


In my system the number is 60, which means that if the DataFrame contains more than 60 rows, the print(df) statement will return only the headers and the first and last 5 rows.

You can change the maximum rows number with the same statement.

In [ ]:
# Increase the maximum number of rows to display the entire data DataFrame:

import pandas as pd

pd.options.display.max_rows = 9999

print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

In [28]:
# Increase the maximum number of rows to display the entire forestfires DataFrame:

# Code Here
import pandas as pd

pd.options.display.max_rows = 9999

print(df1)

     X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain  \
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0   
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0   
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0   
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2   
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0   
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0   
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0   
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0   
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0   
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0   
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0   
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3   38.0   4.0   0.0   
12   6  5   aug  fri  63.

**Pandas - Analyzing DataFrames**

**Viewing the Data**

One of the most used method for getting a quick overview of the DataFrame, is the head() method.

The head() method returns the headers and a specified number of rows, starting from the top.

In [ ]:
# Get a quick overview by printing the first 10 rows of the data DataFrame:

import pandas as pd

print(df.head(10))

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0


In [29]:
# Get a quick overview by printing the first 10 rows of the forestfires DataFrame:

# Code Here
import pandas as pd

print(df1.head(10))

   X  Y month  day  FFMC    DMC     DC   ISI  temp    RH  wind  rain  area
0  7  5   mar  fri  86.2   26.2   94.3   5.1   8.2  51.0   6.7   0.0   0.0
1  7  4   oct  tue  90.6   35.4  669.1   6.7  18.0  33.0   0.9   0.0   0.0
2  7  4   oct  sat  90.6   43.7  686.9   6.7  14.6  33.0   1.3   0.0   0.0
3  8  6   mar  fri  91.7   33.3   77.5   9.0   8.3  97.0   4.0   0.2   0.0
4  8  6   mar  sun  89.3   51.3  102.2   9.6  11.4  99.0   1.8   0.0   0.0
5  8  6   aug  sun  92.3   85.3  488.0  14.7  22.2  29.0   5.4   0.0   0.0
6  8  6   aug  mon  92.3   88.9  495.6   8.5  24.1  27.0   3.1   0.0   0.0
7  8  6   aug  mon  91.5  145.4  608.2  10.7   8.0  86.0   2.2   0.0   0.0
8  8  6   sep  tue  91.0  129.5  692.6   7.0  13.1  63.0   5.4   0.0   0.0
9  7  5   sep  sat  92.5   88.0  698.6   7.1  22.8  40.0   4.0   0.0   0.0


In [ ]:
# Print the first 5 rows of the data DataFrame:

import pandas as pd

print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


In [30]:
# Print the first 5 rows of the forestfires DataFrame:

# Code Here
import pandas as pd

print(df1.head())

   X  Y month  day  FFMC   DMC     DC  ISI  temp    RH  wind  rain  area
0  7  5   mar  fri  86.2  26.2   94.3  5.1   8.2  51.0   6.7   0.0   0.0
1  7  4   oct  tue  90.6  35.4  669.1  6.7  18.0  33.0   0.9   0.0   0.0
2  7  4   oct  sat  90.6  43.7  686.9  6.7  14.6  33.0   1.3   0.0   0.0
3  8  6   mar  fri  91.7  33.3   77.5  9.0   8.3  97.0   4.0   0.2   0.0
4  8  6   mar  sun  89.3  51.3  102.2  9.6  11.4  99.0   1.8   0.0   0.0


There is also a tail() method for viewing the last rows of the DataFrame.

The tail() method returns the headers and a specified number of rows, starting from the bottom.

In [ ]:
# Print the last 5 rows of the data DataFrame:

print(df.tail())

     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


In [31]:
# Print the last 5 rows of the forestfires DataFrame:

# Code Here
print(df1.tail())

     X  Y month  day  FFMC    DMC     DC   ISI  temp    RH  wind  rain   area
512  4  3   aug  sun  81.6   56.7  665.6   1.9  27.8  32.0   2.7   0.0   6.44
513  2  4   aug  sun  81.6   56.7  665.6   1.9  21.9  71.0   5.8   0.0  54.29
514  7  4   aug  sun  81.6   56.7  665.6   1.9  21.2  70.0   6.7   0.0  11.16
515  1  4   aug  sat  94.4  146.0  614.7  11.3  25.6  42.0   4.0   0.0   0.00
516  6  3   nov  tue  79.5    3.0  106.7   1.1  11.8  31.0   4.5   0.0   0.00


**Info About the Data**

The DataFrames object has a method called info(), that gives you more information about the data set.

In [ ]:
# Print information about the data:

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


The result tells us there are 169 rows and 4 columns:

And the name of each column, with the data type:

\

**Null Values**

The info() method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are 164 of 169 Non-Null values in the "Calories" column.

Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.

Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values. This is a step towards what is called cleaning data, and you will learn more about that in the next chapters.

In [32]:
# Print information about the forestfires:

# Code Here
print(df1.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 13 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   X       517 non-null    int64  
 1   Y       517 non-null    int64  
 2   month   517 non-null    object 
 3   day     517 non-null    object 
 4   FFMC    515 non-null    float64
 5   DMC     505 non-null    float64
 6   DC      505 non-null    float64
 7   ISI     511 non-null    float64
 8   temp    502 non-null    float64
 9   RH      506 non-null    float64
 10  wind    513 non-null    float64
 11  rain    515 non-null    float64
 12  area    517 non-null    float64
dtypes: float64(9), int64(2), object(2)
memory usage: 52.6+ KB
None



\
**Pandas - Cleaning Data**

**Data Cleaning**

Data cleaning means fixing bad data in your data set.

Bad data could be:

Empty cells
Data in wrong format
Wrong data
Duplicates


**Pandas - Cleaning Empty Cells**

**Empty Cells**

Empty cells can potentially give you a wrong result when you analyze data.

**Remove Rows**

One way to deal with empty cells is to remove rows that contain empty cells.

This is usually OK, since data sets can be very big, and removing a few rows will not have a big impact on the result.

In [ ]:
# Return a new Data Frame with no empty cells:

import pandas as pd

df = pd.read_csv('data.csv')

new_df = df.dropna()

print(new_df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

In [33]:
# Return a forestfires new Data Frame with no empty cells:

# Code Here
import pandas as pd

df1 = pd.read_csv('forestfires.csv')

new_df1 = df1.dropna()

print(new_df1.to_string())

     X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain     area
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0     0.00
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0     0.00
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0     0.00
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2     0.00
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0     0.00
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0     0.00
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0     0.00
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0     0.00
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0     0.00
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0     0.00
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0     0.00
11   7  5   sep  sat  92.8  

Note: By default, the dropna() method returns a new DataFrame, and will not change the original.

If you want to change the original DataFrame, use the inplace = True argument:

In [ ]:
# Remove all rows with NULL values:

import pandas as pd

df = pd.read_csv('data.csv')

df.dropna(inplace = True)

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

In [34]:
# Remove all rows with NULL values (forestfires.csv):

import pandas as pd

forestfires = pd.read_csv('forestfires.csv')

# Code Here
df1.dropna(inplace = True)

print(df1.to_string())

     X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain     area
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0     0.00
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0     0.00
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0     0.00
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2     0.00
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0     0.00
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0     0.00
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0     0.00
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0     0.00
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0     0.00
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0     0.00
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0     0.00
11   7  5   sep  sat  92.8  

Note: Now, the dropna(inplace = True) will NOT return a new DataFrame, but it will remove all rows containing NULL values from the original DataFrame.

\

**Replace Empty Values**

Another way of dealing with empty cells is to insert a new value instead.

This way you do not have to delete entire rows just because of some empty cells.

The fillna() method allows us to replace empty cells with a value:

In [ ]:
# Replace NULL values with the number 130:

import pandas as pd

df = pd.read_csv('data.csv')

df.fillna(130, inplace = True)

df

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     130.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60    102       126     334.5
25         60    100       120     250.0
26         60     92       118     241.0
27         60    103       132     130.0
28         60    100       132     280.0
29         60    102       129     380.3
30         60     92       115     243.0
31         45     90       112     180.1
32         60    101       124     299.0
33         60     93       113     223.0
34         60    107       136     361.0
35         60    114       140     415.0
36         60    102       127     300.0
37         60    100       120     300.0
38         60    100       120     300.0
39         45    104       129     266.0
40         45     90       112     180.1
41         60     98       126     286.0
42         60    100       122     329.4
43         60    111       138     400.0
44         60    111       131     397.0
45         60     99       119     273.0
46         60    109       153     387.6
47         45    111       136     300.0
48         45    108       129     298.0
49         60    111       139     397.6
50         60    107       136     380.2
51         80    123       146     643.1
52         60    106       130     263.0
53         60    118       151     486.0
54         30    136       175     238.0
55         60    121       146     450.7
56         60    118       121     413.0
57         45    115       144     305.0
58         20    153       172     226.4
59         45    123       152     321.0
60        210    108       160    1376.0
61        160    110       137    1034.4
62        160    109       135     853.0
63         45    118       141     341.0
64         20    110       130     131.4
65        180     90       130     800.4
66        150    105       135     873.4
67        150    107       130     816.0
68         20    106       136     110.4
69        300    108       143    1500.2
70        150     97       129    1115.0
71         60    109       153     387.6
72         90    100       127     700.0
73        150     97       127     953.2
74         45    114       146     304.0
75         90     98       125     563.2
76         45    105       134     251.0
77         45    110       141     300.0
78        120    100       130     500.4
79        270    100       131    1729.0
80         30    159       182     319.2
81         45    149       169     344.0
82         30    103       139     151.1
83        120    100       130     500.0
84         45    100       120     225.3
85         30    151       170     300.0
86         45    102       136     234.0
87        120    100       157    1000.1
88         45    129       103     242.0
89         20     83       107      50.3
90        180    101       127     600.1
91         45    107       137     130.0
92         30     90       107     105.3
93         15     80       100      50.5
94         20    150       171     127.4
95         20    151       168     229.4
96         30     95       1

In [35]:
# Replace NULL values with the number 74 (forestfires.csv):

# Code Here
df1.fillna(74, inplace = True)

df1

X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain  \
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0   
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0   
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0   
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2   
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0   
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0   
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0   
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0   
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0   
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0   
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0   
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3   38.0   4.0   0.0   
13   6  5   sep  mon  90.9  126.5  686.5   7.0  21.3   42.0   2.2   0.0   
14   6  5   sep  wed  92.9  133.3  699.6   9.2  26.4   21.0   4.5   0.0   
15   6  5   sep  fri  93.3  141.2  713.9  13.9  22.9   44.0   5.4   0.0   
17   8  5   oct  mon  84.9   32.8  664.2   3.0  16.7   47.0   4.9   0.0   
18   6  4   mar  wed  89.2   27.9   70.8   6.3  15.9   35.0   4.0   0.0   
20   6  4   sep  tue  91.0  129.5  692.6   7.0  18.3   40.0   2.7   0.0   
21   5  4   sep  mon  91.8   78.5  724.3   9.2  19.1   38.0   2.7   0.0   
22   7  4   jun  sun  94.3   96.3  200.0  56.1  21.0   44.0   4.5   0.0   
24   7  4   aug  sat  93.5  139.4  594.2  20.3  23.7   32.0   5.8   0.0   
25   7  4   aug  sun  91.4  142.4  601.4  10.6  16.3   60.0   5.4   0.0   
26   7  4   sep  fri  92.4  117.9  668.0  12.2  19.0   34.0   5.8   0.0   
28   6  3   sep  sat  93.4  145.4  721.4   8.1  30.2   24.0   2.7   0.0   
29   6  3   sep  sun  93.5  149.3  728.6   8.1  22.8   39.0   3.6   0.0   
31   6  3   sep  mon  88.6   91.8  709.9   7.1  11.2   78.0   7.6   0.0   
32   6  3   sep  fri  88.6   69.7  706.8   5.8  20.6   37.0   1.8   0.0   
33   6  3   sep  sun  91.7   75.6  718.3   7.8  17.7   39.0   3.6   0.0   
35   6  3   sep  tue  90.3   80.7  730.2   6.3  18.2   62.0   4.5   0.0   
36   6  3   oct  tue  90.6   35.4  669.1   6.7  21.7   24.0   4.5   0.0   
37   7  4   oct  fri  90.0   41.5  682.6   8.7  11.3   60.0   5.4   0.0   
38   7  3   oct  sat  90.6   43.7  686.9   6.7  17.8   27.0   4.0   0.0   
39   4  4   mar  tue  88.1   25.7   67.6   3.8  14.1   43.0   2.7   0.0   
40   4  4   jul  tue  79.5   60.6  366.7   1.5  23.3   37.0   3.1   0.0   
41   4  4   aug  sat  90.2   96.9  624.2   8.9  18.4   42.0   6.7   0.0   
43   4  4   sep  sat  92.5   88.0  698.6   7.1  19.6   48.0   2.7   0.0   
45   5  6   sep  wed  94.3   85.1  692.3  15.9  25.9   24.0   4.0   0.0   
46   5  6   sep  mon  90.9  126.5  686.5   7.0  14.7   70.0   3.6   0.0   
48   4  4   mar  mon  87.2   23.9   64.7   4.1  11.8   35.0   1.8   0.0   
52   4  3   aug  wed  92.1  111.2  654.1   9.6  20.4   42.0   4.9   0.0   
53   4  3   aug  wed  92.1  111.2  654.1   9.6  20.4   42.0   4.9   0.0   
54   4  3   aug  thu  91.7  114.3  661.3   6.3  17.6   45.0   3.6   0.0   
55   4  3   sep  thu  92.9  137.0  706.4   9.2  27.7   24.0   2.2   0.0   
56   4  3   sep  tue  90.3   80.7  730.2   6.3  17.8   63.0   4.9   0.0   
57   4  3   oct  sun  92.6   46.5  691.8   8.8  13.8   50.0   2.7   0.0   
58   2  2   feb  mon  84.0    9.3   34.0   2.1  13.9   40.0   5.4   0.0   
59   2  2   feb  fri  86.6   13.2   43.0   5.3  12.3   51.0   0.9   0.0   
60   2  2   mar  sun  89.3   51.3  102.2   9.6  11.5   39.0   5.8   0.0   
61   2  2   mar  sun  89.3   51.3  102.2   9.6   5.5   59.0   6.3   0.0   
62   2  2   aug  thu  93.0   75.3  466.6   7.7  18.8   35.0   4.9   0.0   
64   2  2   aug  mon  91.1  103.2  638.8   5.8  23.1   31.0   3.1   0.0   
67   2  2   sep  fri  92.4  117.9  668.0  12.2  19.6   33.0   5.4   0.0   
70   4  5   mar  fri  91.2   4

**Replace Only For Specified Columns**

The example above replaces all empty cells in the whole Data Frame.

To only replace empty values for one column, specify the column name for the DataFrame:

In [ ]:
# Replace NULL values in the "Calories" columns with the number 130:

import pandas as pd

df = pd.read_csv('data.csv')

df["Calories"].fillna(130, inplace = True)

df

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     130.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60    102       126     334.5
25         60    100       120     250.0
26         60     92       118     241.0
27         60    103       132     130.0
28         60    100       132     280.0
29         60    102       129     380.3
30         60     92       115     243.0
31         45     90       112     180.1
32         60    101       124     299.0
33         60     93       113     223.0
34         60    107       136     361.0
35         60    114       140     415.0
36         60    102       127     300.0
37         60    100       120     300.0
38         60    100       120     300.0
39         45    104       129     266.0
40         45     90       112     180.1
41         60     98       126     286.0
42         60    100       122     329.4
43         60    111       138     400.0
44         60    111       131     397.0
45         60     99       119     273.0
46         60    109       153     387.6
47         45    111       136     300.0
48         45    108       129     298.0
49         60    111       139     397.6
50         60    107       136     380.2
51         80    123       146     643.1
52         60    106       130     263.0
53         60    118       151     486.0
54         30    136       175     238.0
55         60    121       146     450.7
56         60    118       121     413.0
57         45    115       144     305.0
58         20    153       172     226.4
59         45    123       152     321.0
60        210    108       160    1376.0
61        160    110       137    1034.4
62        160    109       135     853.0
63         45    118       141     341.0
64         20    110       130     131.4
65        180     90       130     800.4
66        150    105       135     873.4
67        150    107       130     816.0
68         20    106       136     110.4
69        300    108       143    1500.2
70        150     97       129    1115.0
71         60    109       153     387.6
72         90    100       127     700.0
73        150     97       127     953.2
74         45    114       146     304.0
75         90     98       125     563.2
76         45    105       134     251.0
77         45    110       141     300.0
78        120    100       130     500.4
79        270    100       131    1729.0
80         30    159       182     319.2
81         45    149       169     344.0
82         30    103       139     151.1
83        120    100       130     500.0
84         45    100       120     225.3
85         30    151       170     300.0
86         45    102       136     234.0
87        120    100       157    1000.1
88         45    129       103     242.0
89         20     83       107      50.3
90        180    101       127     600.1
91         45    107       137     130.0
92         30     90       107     105.3
93         15     80       100      50.5
94         20    150       171     127.4
95         20    151       168     229.4
96         30     95       1

In [36]:
# Replace NULL values in the (forestfires.csv) "temp" columns with the number 18:

# Code Here
df1["temp"].fillna(18, inplace = True)

df1

X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain  \
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0   
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0   
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0   
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2   
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0   
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0   
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0   
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0   
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0   
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0   
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0   
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3   38.0   4.0   0.0   
13   6  5   sep  mon  90.9  126.5  686.5   7.0  21.3   42.0   2.2   0.0   
14   6  5   sep  wed  92.9  133.3  699.6   9.2  26.4   21.0   4.5   0.0   
15   6  5   sep  fri  93.3  141.2  713.9  13.9  22.9   44.0   5.4   0.0   
17   8  5   oct  mon  84.9   32.8  664.2   3.0  16.7   47.0   4.9   0.0   
18   6  4   mar  wed  89.2   27.9   70.8   6.3  15.9   35.0   4.0   0.0   
20   6  4   sep  tue  91.0  129.5  692.6   7.0  18.3   40.0   2.7   0.0   
21   5  4   sep  mon  91.8   78.5  724.3   9.2  19.1   38.0   2.7   0.0   
22   7  4   jun  sun  94.3   96.3  200.0  56.1  21.0   44.0   4.5   0.0   
24   7  4   aug  sat  93.5  139.4  594.2  20.3  23.7   32.0   5.8   0.0   
25   7  4   aug  sun  91.4  142.4  601.4  10.6  16.3   60.0   5.4   0.0   
26   7  4   sep  fri  92.4  117.9  668.0  12.2  19.0   34.0   5.8   0.0   
28   6  3   sep  sat  93.4  145.4  721.4   8.1  30.2   24.0   2.7   0.0   
29   6  3   sep  sun  93.5  149.3  728.6   8.1  22.8   39.0   3.6   0.0   
31   6  3   sep  mon  88.6   91.8  709.9   7.1  11.2   78.0   7.6   0.0   
32   6  3   sep  fri  88.6   69.7  706.8   5.8  20.6   37.0   1.8   0.0   
33   6  3   sep  sun  91.7   75.6  718.3   7.8  17.7   39.0   3.6   0.0   
35   6  3   sep  tue  90.3   80.7  730.2   6.3  18.2   62.0   4.5   0.0   
36   6  3   oct  tue  90.6   35.4  669.1   6.7  21.7   24.0   4.5   0.0   
37   7  4   oct  fri  90.0   41.5  682.6   8.7  11.3   60.0   5.4   0.0   
38   7  3   oct  sat  90.6   43.7  686.9   6.7  17.8   27.0   4.0   0.0   
39   4  4   mar  tue  88.1   25.7   67.6   3.8  14.1   43.0   2.7   0.0   
40   4  4   jul  tue  79.5   60.6  366.7   1.5  23.3   37.0   3.1   0.0   
41   4  4   aug  sat  90.2   96.9  624.2   8.9  18.4   42.0   6.7   0.0   
43   4  4   sep  sat  92.5   88.0  698.6   7.1  19.6   48.0   2.7   0.0   
45   5  6   sep  wed  94.3   85.1  692.3  15.9  25.9   24.0   4.0   0.0   
46   5  6   sep  mon  90.9  126.5  686.5   7.0  14.7   70.0   3.6   0.0   
48   4  4   mar  mon  87.2   23.9   64.7   4.1  11.8   35.0   1.8   0.0   
52   4  3   aug  wed  92.1  111.2  654.1   9.6  20.4   42.0   4.9   0.0   
53   4  3   aug  wed  92.1  111.2  654.1   9.6  20.4   42.0   4.9   0.0   
54   4  3   aug  thu  91.7  114.3  661.3   6.3  17.6   45.0   3.6   0.0   
55   4  3   sep  thu  92.9  137.0  706.4   9.2  27.7   24.0   2.2   0.0   
56   4  3   sep  tue  90.3   80.7  730.2   6.3  17.8   63.0   4.9   0.0   
57   4  3   oct  sun  92.6   46.5  691.8   8.8  13.8   50.0   2.7   0.0   
58   2  2   feb  mon  84.0    9.3   34.0   2.1  13.9   40.0   5.4   0.0   
59   2  2   feb  fri  86.6   13.2   43.0   5.3  12.3   51.0   0.9   0.0   
60   2  2   mar  sun  89.3   51.3  102.2   9.6  11.5   39.0   5.8   0.0   
61   2  2   mar  sun  89.3   51.3  102.2   9.6   5.5   59.0   6.3   0.0   
62   2  2   aug  thu  93.0   75.3  466.6   7.7  18.8   35.0   4.9   0.0   
64   2  2   aug  mon  91.1  103.2  638.8   5.8  23.1   31.0   3.1   0.0   
67   2  2   sep  fri  92.4  117.9  668.0  12.2  19.6   33.0   5.4   0.0   
70   4  5   mar  fri  91.2   4

**Replace Using Mean, Median, or Mode**

A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

In [ ]:
# Calculate the MEAN, and replace any empty values with it:

import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].mean()

df["Calories"].fillna(x, inplace = True)

df

Duration  Pulse  Maxpulse     Calories
0          60    110       130   409.100000
1          60    117       145   479.000000
2          60    103       135   340.000000
3          45    109       175   282.400000
4          45    117       148   406.000000
5          60    102       127   300.000000
6          60    110       136   374.000000
7          45    104       134   253.300000
8          30    109       133   195.100000
9          60     98       124   269.000000
10         60    103       147   329.300000
11         60    100       120   250.700000
12         60    106       128   345.300000
13         60    104       132   379.300000
14         60     98       123   275.000000
15         60     98       120   215.200000
16         60    100       120   300.000000
17         45     90       112   375.790244
18         60    103       123   323.000000
19         45     97       125   243.000000
20         60    108       131   364.200000
21         45    100       119   282.000000
22         60    130       101   300.000000
23         45    105       132   246.000000
24         60    102       126   334.500000
25         60    100       120   250.000000
26         60     92       118   241.000000
27         60    103       132   375.790244
28         60    100       132   280.000000
29         60    102       129   380.300000
30         60     92       115   243.000000
31         45     90       112   180.100000
32         60    101       124   299.000000
33         60     93       113   223.000000
34         60    107       136   361.000000
35         60    114       140   415.000000
36         60    102       127   300.000000
37         60    100       120   300.000000
38         60    100       120   300.000000
39         45    104       129   266.000000
40         45     90       112   180.100000
41         60     98       126   286.000000
42         60    100       122   329.400000
43         60    111       138   400.000000
44         60    111       131   397.000000
45         60     99       119   273.000000
46         60    109       153   387.600000
47         45    111       136   300.000000
48         45    108       129   298.000000
49         60    111       139   397.600000
50         60    107       136   380.200000
51         80    123       146   643.100000
52         60    106       130   263.000000
53         60    118       151   486.000000
54         30    136       175   238.000000
55         60    121       146   450.700000
56         60    118       121   413.000000
57         45    115       144   305.000000
58         20    153       172   226.400000
59         45    123       152   321.000000
60        210    108       160  1376.000000
61        160    110       137  1034.400000
62        160    109       135   853.000000
63         45    118       141   341.000000
64         20    110       130   131.400000
65        180     90       130   800.400000
66        150    105       135   873.400000
67        150    107       130   816.000000
68         20    106       136   110.400000
69        300    108       143  1500.200000
70        150     97       129  1115.000000
71         60    109       153   387.600000
72         90    100       127   700.000000
73        150     97       127   953.200000
74         45    114       146   304.000000
75         90     98       125   563.200000
76         45    105       134   251.000000
77         45    110       141   300.000000
78        120    100       130   500.400000
79        270    100       131  1729.000000
80         30    159       182   319.200000
81         45    149       169   344.000000
82         30    103       139   151.100000
83        120    100       130   500.000000
84         45    100       120   225.300000
85         30    151       170   300.000000
86         45    102       136   234.000000
87        120    100       157  1000.100000
88         45    129       103   242.000000
89         20     83       107    50.300000
9

In [37]:
# Calculate the MEAN, and replace any empty values with it (forestfires) "temp" columns:

# Code Here
y = df1["temp"].mean()

df1["temp"].fillna(y, inplace = True)

df1

X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain  \
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0   
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0   
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0   
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2   
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0   
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0   
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0   
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0   
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0   
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0   
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0   
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3   38.0   4.0   0.0   
13   6  5   sep  mon  90.9  126.5  686.5   7.0  21.3   42.0   2.2   0.0   
14   6  5   sep  wed  92.9  133.3  699.6   9.2  26.4   21.0   4.5   0.0   
15   6  5   sep  fri  93.3  141.2  713.9  13.9  22.9   44.0   5.4   0.0   
17   8  5   oct  mon  84.9   32.8  664.2   3.0  16.7   47.0   4.9   0.0   
18   6  4   mar  wed  89.2   27.9   70.8   6.3  15.9   35.0   4.0   0.0   
20   6  4   sep  tue  91.0  129.5  692.6   7.0  18.3   40.0   2.7   0.0   
21   5  4   sep  mon  91.8   78.5  724.3   9.2  19.1   38.0   2.7   0.0   
22   7  4   jun  sun  94.3   96.3  200.0  56.1  21.0   44.0   4.5   0.0   
24   7  4   aug  sat  93.5  139.4  594.2  20.3  23.7   32.0   5.8   0.0   
25   7  4   aug  sun  91.4  142.4  601.4  10.6  16.3   60.0   5.4   0.0   
26   7  4   sep  fri  92.4  117.9  668.0  12.2  19.0   34.0   5.8   0.0   
28   6  3   sep  sat  93.4  145.4  721.4   8.1  30.2   24.0   2.7   0.0   
29   6  3   sep  sun  93.5  149.3  728.6   8.1  22.8   39.0   3.6   0.0   
31   6  3   sep  mon  88.6   91.8  709.9   7.1  11.2   78.0   7.6   0.0   
32   6  3   sep  fri  88.6   69.7  706.8   5.8  20.6   37.0   1.8   0.0   
33   6  3   sep  sun  91.7   75.6  718.3   7.8  17.7   39.0   3.6   0.0   
35   6  3   sep  tue  90.3   80.7  730.2   6.3  18.2   62.0   4.5   0.0   
36   6  3   oct  tue  90.6   35.4  669.1   6.7  21.7   24.0   4.5   0.0   
37   7  4   oct  fri  90.0   41.5  682.6   8.7  11.3   60.0   5.4   0.0   
38   7  3   oct  sat  90.6   43.7  686.9   6.7  17.8   27.0   4.0   0.0   
39   4  4   mar  tue  88.1   25.7   67.6   3.8  14.1   43.0   2.7   0.0   
40   4  4   jul  tue  79.5   60.6  366.7   1.5  23.3   37.0   3.1   0.0   
41   4  4   aug  sat  90.2   96.9  624.2   8.9  18.4   42.0   6.7   0.0   
43   4  4   sep  sat  92.5   88.0  698.6   7.1  19.6   48.0   2.7   0.0   
45   5  6   sep  wed  94.3   85.1  692.3  15.9  25.9   24.0   4.0   0.0   
46   5  6   sep  mon  90.9  126.5  686.5   7.0  14.7   70.0   3.6   0.0   
48   4  4   mar  mon  87.2   23.9   64.7   4.1  11.8   35.0   1.8   0.0   
52   4  3   aug  wed  92.1  111.2  654.1   9.6  20.4   42.0   4.9   0.0   
53   4  3   aug  wed  92.1  111.2  654.1   9.6  20.4   42.0   4.9   0.0   
54   4  3   aug  thu  91.7  114.3  661.3   6.3  17.6   45.0   3.6   0.0   
55   4  3   sep  thu  92.9  137.0  706.4   9.2  27.7   24.0   2.2   0.0   
56   4  3   sep  tue  90.3   80.7  730.2   6.3  17.8   63.0   4.9   0.0   
57   4  3   oct  sun  92.6   46.5  691.8   8.8  13.8   50.0   2.7   0.0   
58   2  2   feb  mon  84.0    9.3   34.0   2.1  13.9   40.0   5.4   0.0   
59   2  2   feb  fri  86.6   13.2   43.0   5.3  12.3   51.0   0.9   0.0   
60   2  2   mar  sun  89.3   51.3  102.2   9.6  11.5   39.0   5.8   0.0   
61   2  2   mar  sun  89.3   51.3  102.2   9.6   5.5   59.0   6.3   0.0   
62   2  2   aug  thu  93.0   75.3  466.6   7.7  18.8   35.0   4.9   0.0   
64   2  2   aug  mon  91.1  103.2  638.8   5.8  23.1   31.0   3.1   0.0   
67   2  2   sep  fri  92.4  117.9  668.0  12.2  19.6   33.0   5.4   0.0   
70   4  5   mar  fri  91.2   4

Mean = the average value (the sum of all values divided by number of values).

In [ ]:
# Calculate the MEDIAN, and replace any empty values with it:

import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].median()

df["Calories"].fillna(x, inplace = True)

df

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     318.6
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60    102       126     334.5
25         60    100       120     250.0
26         60     92       118     241.0
27         60    103       132     318.6
28         60    100       132     280.0
29         60    102       129     380.3
30         60     92       115     243.0
31         45     90       112     180.1
32         60    101       124     299.0
33         60     93       113     223.0
34         60    107       136     361.0
35         60    114       140     415.0
36         60    102       127     300.0
37         60    100       120     300.0
38         60    100       120     300.0
39         45    104       129     266.0
40         45     90       112     180.1
41         60     98       126     286.0
42         60    100       122     329.4
43         60    111       138     400.0
44         60    111       131     397.0
45         60     99       119     273.0
46         60    109       153     387.6
47         45    111       136     300.0
48         45    108       129     298.0
49         60    111       139     397.6
50         60    107       136     380.2
51         80    123       146     643.1
52         60    106       130     263.0
53         60    118       151     486.0
54         30    136       175     238.0
55         60    121       146     450.7
56         60    118       121     413.0
57         45    115       144     305.0
58         20    153       172     226.4
59         45    123       152     321.0
60        210    108       160    1376.0
61        160    110       137    1034.4
62        160    109       135     853.0
63         45    118       141     341.0
64         20    110       130     131.4
65        180     90       130     800.4
66        150    105       135     873.4
67        150    107       130     816.0
68         20    106       136     110.4
69        300    108       143    1500.2
70        150     97       129    1115.0
71         60    109       153     387.6
72         90    100       127     700.0
73        150     97       127     953.2
74         45    114       146     304.0
75         90     98       125     563.2
76         45    105       134     251.0
77         45    110       141     300.0
78        120    100       130     500.4
79        270    100       131    1729.0
80         30    159       182     319.2
81         45    149       169     344.0
82         30    103       139     151.1
83        120    100       130     500.0
84         45    100       120     225.3
85         30    151       170     300.0
86         45    102       136     234.0
87        120    100       157    1000.1
88         45    129       103     242.0
89         20     83       107      50.3
90        180    101       127     600.1
91         45    107       137     318.6
92         30     90       107     105.3
93         15     80       100      50.5
94         20    150       171     127.4
95         20    151       168     229.4
96         30     95       1

In [38]:
# Calculate the MEDIAN, and replace any empty values with it (forestfires) "temp" columns:

# Code Here
y = df1["temp"].median()

df1["temp"].fillna(y, inplace = True)

df1

X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain  \
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0   
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0   
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0   
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2   
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0   
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0   
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0   
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0   
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0   
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0   
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0   
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3   38.0   4.0   0.0   
13   6  5   sep  mon  90.9  126.5  686.5   7.0  21.3   42.0   2.2   0.0   
14   6  5   sep  wed  92.9  133.3  699.6   9.2  26.4   21.0   4.5   0.0   
15   6  5   sep  fri  93.3  141.2  713.9  13.9  22.9   44.0   5.4   0.0   
17   8  5   oct  mon  84.9   32.8  664.2   3.0  16.7   47.0   4.9   0.0   
18   6  4   mar  wed  89.2   27.9   70.8   6.3  15.9   35.0   4.0   0.0   
20   6  4   sep  tue  91.0  129.5  692.6   7.0  18.3   40.0   2.7   0.0   
21   5  4   sep  mon  91.8   78.5  724.3   9.2  19.1   38.0   2.7   0.0   
22   7  4   jun  sun  94.3   96.3  200.0  56.1  21.0   44.0   4.5   0.0   
24   7  4   aug  sat  93.5  139.4  594.2  20.3  23.7   32.0   5.8   0.0   
25   7  4   aug  sun  91.4  142.4  601.4  10.6  16.3   60.0   5.4   0.0   
26   7  4   sep  fri  92.4  117.9  668.0  12.2  19.0   34.0   5.8   0.0   
28   6  3   sep  sat  93.4  145.4  721.4   8.1  30.2   24.0   2.7   0.0   
29   6  3   sep  sun  93.5  149.3  728.6   8.1  22.8   39.0   3.6   0.0   
31   6  3   sep  mon  88.6   91.8  709.9   7.1  11.2   78.0   7.6   0.0   
32   6  3   sep  fri  88.6   69.7  706.8   5.8  20.6   37.0   1.8   0.0   
33   6  3   sep  sun  91.7   75.6  718.3   7.8  17.7   39.0   3.6   0.0   
35   6  3   sep  tue  90.3   80.7  730.2   6.3  18.2   62.0   4.5   0.0   
36   6  3   oct  tue  90.6   35.4  669.1   6.7  21.7   24.0   4.5   0.0   
37   7  4   oct  fri  90.0   41.5  682.6   8.7  11.3   60.0   5.4   0.0   
38   7  3   oct  sat  90.6   43.7  686.9   6.7  17.8   27.0   4.0   0.0   
39   4  4   mar  tue  88.1   25.7   67.6   3.8  14.1   43.0   2.7   0.0   
40   4  4   jul  tue  79.5   60.6  366.7   1.5  23.3   37.0   3.1   0.0   
41   4  4   aug  sat  90.2   96.9  624.2   8.9  18.4   42.0   6.7   0.0   
43   4  4   sep  sat  92.5   88.0  698.6   7.1  19.6   48.0   2.7   0.0   
45   5  6   sep  wed  94.3   85.1  692.3  15.9  25.9   24.0   4.0   0.0   
46   5  6   sep  mon  90.9  126.5  686.5   7.0  14.7   70.0   3.6   0.0   
48   4  4   mar  mon  87.2   23.9   64.7   4.1  11.8   35.0   1.8   0.0   
52   4  3   aug  wed  92.1  111.2  654.1   9.6  20.4   42.0   4.9   0.0   
53   4  3   aug  wed  92.1  111.2  654.1   9.6  20.4   42.0   4.9   0.0   
54   4  3   aug  thu  91.7  114.3  661.3   6.3  17.6   45.0   3.6   0.0   
55   4  3   sep  thu  92.9  137.0  706.4   9.2  27.7   24.0   2.2   0.0   
56   4  3   sep  tue  90.3   80.7  730.2   6.3  17.8   63.0   4.9   0.0   
57   4  3   oct  sun  92.6   46.5  691.8   8.8  13.8   50.0   2.7   0.0   
58   2  2   feb  mon  84.0    9.3   34.0   2.1  13.9   40.0   5.4   0.0   
59   2  2   feb  fri  86.6   13.2   43.0   5.3  12.3   51.0   0.9   0.0   
60   2  2   mar  sun  89.3   51.3  102.2   9.6  11.5   39.0   5.8   0.0   
61   2  2   mar  sun  89.3   51.3  102.2   9.6   5.5   59.0   6.3   0.0   
62   2  2   aug  thu  93.0   75.3  466.6   7.7  18.8   35.0   4.9   0.0   
64   2  2   aug  mon  91.1  103.2  638.8   5.8  23.1   31.0   3.1   0.0   
67   2  2   sep  fri  92.4  117.9  668.0  12.2  19.6   33.0   5.4   0.0   
70   4  5   mar  fri  91.2   4

Median = the value in the middle, after you have sorted all values ascending.

In [ ]:
# Calculate the MODE, and replace any empty values with it:

import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].mode()[0]

df["Calories"].fillna(x, inplace = True)

df

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60    102       126     334.5
25         60    100       120     250.0
26         60     92       118     241.0
27         60    103       132     300.0
28         60    100       132     280.0
29         60    102       129     380.3
30         60     92       115     243.0
31         45     90       112     180.1
32         60    101       124     299.0
33         60     93       113     223.0
34         60    107       136     361.0
35         60    114       140     415.0
36         60    102       127     300.0
37         60    100       120     300.0
38         60    100       120     300.0
39         45    104       129     266.0
40         45     90       112     180.1
41         60     98       126     286.0
42         60    100       122     329.4
43         60    111       138     400.0
44         60    111       131     397.0
45         60     99       119     273.0
46         60    109       153     387.6
47         45    111       136     300.0
48         45    108       129     298.0
49         60    111       139     397.6
50         60    107       136     380.2
51         80    123       146     643.1
52         60    106       130     263.0
53         60    118       151     486.0
54         30    136       175     238.0
55         60    121       146     450.7
56         60    118       121     413.0
57         45    115       144     305.0
58         20    153       172     226.4
59         45    123       152     321.0
60        210    108       160    1376.0
61        160    110       137    1034.4
62        160    109       135     853.0
63         45    118       141     341.0
64         20    110       130     131.4
65        180     90       130     800.4
66        150    105       135     873.4
67        150    107       130     816.0
68         20    106       136     110.4
69        300    108       143    1500.2
70        150     97       129    1115.0
71         60    109       153     387.6
72         90    100       127     700.0
73        150     97       127     953.2
74         45    114       146     304.0
75         90     98       125     563.2
76         45    105       134     251.0
77         45    110       141     300.0
78        120    100       130     500.4
79        270    100       131    1729.0
80         30    159       182     319.2
81         45    149       169     344.0
82         30    103       139     151.1
83        120    100       130     500.0
84         45    100       120     225.3
85         30    151       170     300.0
86         45    102       136     234.0
87        120    100       157    1000.1
88         45    129       103     242.0
89         20     83       107      50.3
90        180    101       127     600.1
91         45    107       137     300.0
92         30     90       107     105.3
93         15     80       100      50.5
94         20    150       171     127.4
95         20    151       168     229.4
96         30     95       1

In [41]:
# Calculate the MODE, and replace any empty values with it (forestfires) "temp" columns:

# Code Here
y = df1["temp"].mode()[0]

df1["temp"].fillna(y, inplace = True)

df1

X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain  \
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0   
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0   
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0   
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2   
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0   
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0   
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0   
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0   
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0   
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0   
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0   
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3   38.0   4.0   0.0   
13   6  5   sep  mon  90.9  126.5  686.5   7.0  21.3   42.0   2.2   0.0   
14   6  5   sep  wed  92.9  133.3  699.6   9.2  26.4   21.0   4.5   0.0   
15   6  5   sep  fri  93.3  141.2  713.9  13.9  22.9   44.0   5.4   0.0   
17   8  5   oct  mon  84.9   32.8  664.2   3.0  16.7   47.0   4.9   0.0   
18   6  4   mar  wed  89.2   27.9   70.8   6.3  15.9   35.0   4.0   0.0   
20   6  4   sep  tue  91.0  129.5  692.6   7.0  18.3   40.0   2.7   0.0   
21   5  4   sep  mon  91.8   78.5  724.3   9.2  19.1   38.0   2.7   0.0   
22   7  4   jun  sun  94.3   96.3  200.0  56.1  21.0   44.0   4.5   0.0   
24   7  4   aug  sat  93.5  139.4  594.2  20.3  23.7   32.0   5.8   0.0   
25   7  4   aug  sun  91.4  142.4  601.4  10.6  16.3   60.0   5.4   0.0   
26   7  4   sep  fri  92.4  117.9  668.0  12.2  19.0   34.0   5.8   0.0   
28   6  3   sep  sat  93.4  145.4  721.4   8.1  30.2   24.0   2.7   0.0   
29   6  3   sep  sun  93.5  149.3  728.6   8.1  22.8   39.0   3.6   0.0   
31   6  3   sep  mon  88.6   91.8  709.9   7.1  11.2   78.0   7.6   0.0   
32   6  3   sep  fri  88.6   69.7  706.8   5.8  20.6   37.0   1.8   0.0   
33   6  3   sep  sun  91.7   75.6  718.3   7.8  17.7   39.0   3.6   0.0   
35   6  3   sep  tue  90.3   80.7  730.2   6.3  18.2   62.0   4.5   0.0   
36   6  3   oct  tue  90.6   35.4  669.1   6.7  21.7   24.0   4.5   0.0   
37   7  4   oct  fri  90.0   41.5  682.6   8.7  11.3   60.0   5.4   0.0   
38   7  3   oct  sat  90.6   43.7  686.9   6.7  17.8   27.0   4.0   0.0   
39   4  4   mar  tue  88.1   25.7   67.6   3.8  14.1   43.0   2.7   0.0   
40   4  4   jul  tue  79.5   60.6  366.7   1.5  23.3   37.0   3.1   0.0   
41   4  4   aug  sat  90.2   96.9  624.2   8.9  18.4   42.0   6.7   0.0   
43   4  4   sep  sat  92.5   88.0  698.6   7.1  19.6   48.0   2.7   0.0   
45   5  6   sep  wed  94.3   85.1  692.3  15.9  25.9   24.0   4.0   0.0   
46   5  6   sep  mon  90.9  126.5  686.5   7.0  14.7   70.0   3.6   0.0   
48   4  4   mar  mon  87.2   23.9   64.7   4.1  11.8   35.0   1.8   0.0   
52   4  3   aug  wed  92.1  111.2  654.1   9.6  20.4   42.0   4.9   0.0   
53   4  3   aug  wed  92.1  111.2  654.1   9.6  20.4   42.0   4.9   0.0   
54   4  3   aug  thu  91.7  114.3  661.3   6.3  17.6   45.0   3.6   0.0   
55   4  3   sep  thu  92.9  137.0  706.4   9.2  27.7   24.0   2.2   0.0   
56   4  3   sep  tue  90.3   80.7  730.2   6.3  17.8   63.0   4.9   0.0   
57   4  3   oct  sun  92.6   46.5  691.8   8.8  13.8   50.0   2.7   0.0   
58   2  2   feb  mon  84.0    9.3   34.0   2.1  13.9   40.0   5.4   0.0   
59   2  2   feb  fri  86.6   13.2   43.0   5.3  12.3   51.0   0.9   0.0   
60   2  2   mar  sun  89.3   51.3  102.2   9.6  11.5   39.0   5.8   0.0   
61   2  2   mar  sun  89.3   51.3  102.2   9.6   5.5   59.0   6.3   0.0   
62   2  2   aug  thu  93.0   75.3  466.6   7.7  18.8   35.0   4.9   0.0   
64   2  2   aug  mon  91.1  103.2  638.8   5.8  23.1   31.0   3.1   0.0   
67   2  2   sep  fri  92.4  117.9  668.0  12.2  19.6   33.0   5.4   0.0   
70   4  5   mar  fri  91.2   4

Mode = the value that appears most frequently.

\
**Pandas - Cleaning Data of Wrong Format**

**Data of Wrong Format**

Cells with data of wrong format can make it difficult, or even impossible, to analyze data.

To fix it, you have two options: remove the rows, or convert all cells in the columns into the same format.

**Convert Into a Correct Format**

In our Data Frame, we have two cells with the wrong format. Check out row 22 and 26, the 'Date' column should be a string that represents a date:

Let's try to convert all cells in the 'Date' column into dates.

Pandas has a to_datetime() method for this:

We will use 'data_Date.csv' dataset.

In [ ]:
import pandas as pd

df = pd.read_csv('data_Date.csv')
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

In [ ]:
# Convert to date:

import pandas as pd

df = pd.read_csv('data_Date.csv')

df['Date'] = pd.to_datetime(df['Date'])

print(df.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

As you can see from the result, the date in row 26 was fixed, but the empty date in row 22 got a NaT (Not a Time) value, in other words an empty value. One way to deal with empty values is simply removing the entire row.

\

Removing Rows
The result from the converting in the example above gave us a NaT value, which can be handled as a NULL value, and we can remove the row by using the dropna() method.

In [ ]:
# Remove rows with a NULL value in the "Date" column:

df.dropna(subset=['Date'], inplace = True)

df

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

**Pandas - Fixing Wrong Data**

**Wrong Data**

"Wrong data" does not have to be "empty cells" or "wrong format", it can just be wrong, like if someone registered "199" instead of "1.99".

Sometimes you can spot wrong data by looking at the data set, because you have an expectation of what it should be.

If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.

It doesn't have to be wrong, but taking in consideration that this is the data set of someone's workout sessions, we conclude with the fact that this person did not work out in 450 minutes.

How can we fix wrong values, like the one for "Duration" in row 7?



**Replacing Values**

One way to fix wrong values is to replace them with something else.

In our example, it is most likely a typo, and the value should be "45" instead of "450", and we could just insert "45" in row 7:

In [ ]:
# Set "Duration" = 45 in row 7:

df.loc[7, 'Duration'] = 45

df

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

In [42]:
# Set "Duration" = 55 in row 17 and 20:

# Code Here
df.loc[17, 'Duration'] = 55
df.loc[20, 'Duration'] = 55

df

X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain  \
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0   
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0   
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0   
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2   
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0   
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0   
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0   
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0   
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0   
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0   
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0   
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3   38.0   4.0   0.0   
12   6  5   aug  fri  63.5   70.8    NaN   0.8  17.0   72.0   6.7   0.0   
13   6  5   sep  mon  90.9  126.5  686.5   7.0  21.3   42.0   2.2   0.0   
14   6  5   sep  wed  92.9  133.3  699.6   9.2  26.4   21.0   4.5   0.0   
15   6  5   sep  fri  93.3  141.2  713.9  13.9  22.9   44.0   5.4   0.0   
16   5  5   mar  sat  91.7   35.8   80.8   NaN  15.1   27.0   5.4   0.0   
17   8  5   oct  mon  84.9   32.8  664.2   3.0  16.7   47.0   4.9   0.0   
18   6  4   mar  wed  89.2   27.9   70.8   6.3  15.9   35.0   4.0   0.0   
19   6  4   apr  sat  86.3   27.4   97.1   5.1   9.3    NaN   4.5   0.0   
20   6  4   sep  tue  91.0  129.5  692.6   7.0  18.3   40.0   2.7   0.0   
21   5  4   sep  mon  91.8   78.5  724.3   9.2  19.1   38.0   2.7   0.0   
22   7  4   jun  sun  94.3   96.3  200.0  56.1  21.0   44.0   4.5   0.0   
23   7  4   aug  sat  90.2    NaN  537.4   NaN  19.5   43.0   5.8   0.0   
24   7  4   aug  sat  93.5  139.4  594.2  20.3  23.7   32.0   5.8   0.0   
25   7  4   aug  sun  91.4  142.4  601.4  10.6  16.3   60.0   5.4   0.0   
26   7  4   sep  fri  92.4  117.9  668.0  12.2  19.0   34.0   5.8   0.0   
27   7  4   sep  mon  90.9  126.5  686.5   7.0   NaN   48.0   1.3   0.0   
28   6  3   sep  sat  93.4  145.4  721.4   8.1  30.2   24.0   2.7   0.0   
29   6  3   sep  sun  93.5  149.3  728.6   8.1  22.8   39.0   3.6   0.0   
30   6  3   sep  fri  94.3   85.1  692.3  15.9  25.4    NaN   3.6   0.0   
31   6  3   sep  mon  88.6   91.8  709.9   7.1  11.2   78.0   7.6   0.0   
32   6  3   sep  fri  88.6   69.7  706.8   5.8  20.6   37.0   1.8   0.0   
33   6  3   sep  sun  91.7   75.6  718.3   7.8  17.7   39.0   3.6   0.0   
34   6  3   sep  mon  91.8    NaN  724.3   9.2  21.2   32.0   2.7   0.0   
35   6  3   sep  tue  90.3   80.7  730.2   6.3  18.2   62.0   4.5   0.0   
36   6  3   oct  tue  90.6   35.4  669.1   6.7  21.7   24.0   4.5   0.0   
37   7  4   oct  fri  90.0   41.5  682.6   8.7  11.3   60.0   5.4   0.0   
38   7  3   oct  sat  90.6   43.7  686.9   6.7  17.8   27.0   4.0   0.0   
39   4  4   mar  tue  88.1   25.7   67.6   3.8  14.1   43.0   2.7   0.0   
40   4  4   jul  tue  79.5   60.6  366.7   1.5  23.3   37.0   3.1   0.0   
41   4  4   aug  sat  90.2   96.9  624.2   8.9  18.4   42.0   6.7   0.0   
42   4  4   aug  tue  94.8  108.3  647.1  17.0   NaN   54.0   5.4   0.0   
43   4  4   sep  sat  92.5   88.0  698.6   7.1  19.6   48.0   2.7   0.0   
44   4  4   sep  wed  90.1   82.9    NaN   6.2  12.9   74.0   4.9   0.0   
45   5  6   sep  wed  94.3   85.1  692.3  15.9  25.9   24.0   4.0   0.0   
46   5  6   sep  mon  90.9  126.5  686.5   7.0  14.7   70.0   3.6   0.0   
47   6  6   jul  mon  94.2   62.3  442.9  11.0  23.0   36.0   3.1   NaN   
48   4  4   mar  mon  87.2   23.9   64.7   4.1  11.8   35.0   1.8   0.0   
49   4  4   mar  mon  87.6   52.2  103.8   5.0   NaN   46.0   5.8   0.0   
50   4  4   sep  thu  92.9    NaN  706.4   9.2  20.8    NaN   1.3   0.0   
51   4  3   aug  sun  90.2   99.6  631.2   6.3   NaN   34.0   2.2   0.0   
52   4  3   aug  wed  92.1  11

For small data sets you might be able to replace the wrong data one by one, but not for big data sets.

To replace wrong data for larger data sets you can create some rules, e.g. set some boundaries for legal values, and replace any values that are outside of the boundaries.

In [44]:
# Loop through all values in the "Duration" column.

# If the value is higher than 120, set it to 120:

# Code Here
for index, value in df.iterrows():
    if value['Duration'] > 120:
        df.loc[index, 'Duration'] = 120
df

X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain  \
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0   
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0   
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0   
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2   
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0   
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0   
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0   
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0   
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0   
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0   
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0   
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3   38.0   4.0   0.0   
12   6  5   aug  fri  63.5   70.8    NaN   0.8  17.0   72.0   6.7   0.0   
13   6  5   sep  mon  90.9  126.5  686.5   7.0  21.3   42.0   2.2   0.0   
14   6  5   sep  wed  92.9  133.3  699.6   9.2  26.4   21.0   4.5   0.0   
15   6  5   sep  fri  93.3  141.2  713.9  13.9  22.9   44.0   5.4   0.0   
16   5  5   mar  sat  91.7   35.8   80.8   NaN  15.1   27.0   5.4   0.0   
17   8  5   oct  mon  84.9   32.8  664.2   3.0  16.7   47.0   4.9   0.0   
18   6  4   mar  wed  89.2   27.9   70.8   6.3  15.9   35.0   4.0   0.0   
19   6  4   apr  sat  86.3   27.4   97.1   5.1   9.3    NaN   4.5   0.0   
20   6  4   sep  tue  91.0  129.5  692.6   7.0  18.3   40.0   2.7   0.0   
21   5  4   sep  mon  91.8   78.5  724.3   9.2  19.1   38.0   2.7   0.0   
22   7  4   jun  sun  94.3   96.3  200.0  56.1  21.0   44.0   4.5   0.0   
23   7  4   aug  sat  90.2    NaN  537.4   NaN  19.5   43.0   5.8   0.0   
24   7  4   aug  sat  93.5  139.4  594.2  20.3  23.7   32.0   5.8   0.0   
25   7  4   aug  sun  91.4  142.4  601.4  10.6  16.3   60.0   5.4   0.0   
26   7  4   sep  fri  92.4  117.9  668.0  12.2  19.0   34.0   5.8   0.0   
27   7  4   sep  mon  90.9  126.5  686.5   7.0   NaN   48.0   1.3   0.0   
28   6  3   sep  sat  93.4  145.4  721.4   8.1  30.2   24.0   2.7   0.0   
29   6  3   sep  sun  93.5  149.3  728.6   8.1  22.8   39.0   3.6   0.0   
30   6  3   sep  fri  94.3   85.1  692.3  15.9  25.4    NaN   3.6   0.0   
31   6  3   sep  mon  88.6   91.8  709.9   7.1  11.2   78.0   7.6   0.0   
32   6  3   sep  fri  88.6   69.7  706.8   5.8  20.6   37.0   1.8   0.0   
33   6  3   sep  sun  91.7   75.6  718.3   7.8  17.7   39.0   3.6   0.0   
34   6  3   sep  mon  91.8    NaN  724.3   9.2  21.2   32.0   2.7   0.0   
35   6  3   sep  tue  90.3   80.7  730.2   6.3  18.2   62.0   4.5   0.0   
36   6  3   oct  tue  90.6   35.4  669.1   6.7  21.7   24.0   4.5   0.0   
37   7  4   oct  fri  90.0   41.5  682.6   8.7  11.3   60.0   5.4   0.0   
38   7  3   oct  sat  90.6   43.7  686.9   6.7  17.8   27.0   4.0   0.0   
39   4  4   mar  tue  88.1   25.7   67.6   3.8  14.1   43.0   2.7   0.0   
40   4  4   jul  tue  79.5   60.6  366.7   1.5  23.3   37.0   3.1   0.0   
41   4  4   aug  sat  90.2   96.9  624.2   8.9  18.4   42.0   6.7   0.0   
42   4  4   aug  tue  94.8  108.3  647.1  17.0   NaN   54.0   5.4   0.0   
43   4  4   sep  sat  92.5   88.0  698.6   7.1  19.6   48.0   2.7   0.0   
44   4  4   sep  wed  90.1   82.9    NaN   6.2  12.9   74.0   4.9   0.0   
45   5  6   sep  wed  94.3   85.1  692.3  15.9  25.9   24.0   4.0   0.0   
46   5  6   sep  mon  90.9  126.5  686.5   7.0  14.7   70.0   3.6   0.0   
47   6  6   jul  mon  94.2   62.3  442.9  11.0  23.0   36.0   3.1   NaN   
48   4  4   mar  mon  87.2   23.9   64.7   4.1  11.8   35.0   1.8   0.0   
49   4  4   mar  mon  87.6   52.2  103.8   5.0   NaN   46.0   5.8   0.0   
50   4  4   sep  thu  92.9    NaN  706.4   9.2  20.8    NaN   1.3   0.0   
51   4  3   aug  sun  90.2   99.6  631.2   6.3   NaN   34.0   2.2   0.0   
52   4  3   aug  wed  92.1  11

**Removing Rows**

Another way of handling wrong data is to remove the rows that contains wrong data.

This way you do not have to find out what to replace them with, and there is a good chance you do not need them to do your analyses.

In [45]:
# Delete rows where "Duration" is higher than 120:

# Code Here
df = df[df['Duration'] <= 120]
df

X  Y month  day  FFMC    DMC     DC  ISI  temp    RH  wind  rain  area  \
17  8  5   oct  mon  84.9   32.8  664.2  3.0  16.7  47.0   4.9   0.0   0.0   
20  6  4   sep  tue  91.0  129.5  692.6  7.0  18.3  40.0   2.7   0.0   0.0   

    Duration  
17      55.0  
20      55.0

**Pandas - Removing Duplicates**

**Discovering Duplicates**

Duplicate rows are rows that have been registered more than one time.

\

By taking a look at our test data set, we can assume that row 11 and 12 are duplicates.

To discover duplicates, we can use the duplicated() method.

The duplicated() method returns a Boolean values for each row:

In [ ]:
# Returns True for every row that is a duplicate, othwerwise False:

print(df.duplicated())

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
dtype: bool


In [46]:
# Returns True for every row that is a duplicate, othwerwise False (forestfires.csv):

# Code Here
print(df1.duplicated())

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
10     False
11     False
13     False
14     False
15     False
17     False
18     False
20     False
21     False
22     False
24     False
25     False
26     False
28     False
29     False
31     False
32     False
33     False
35     False
36     False
37     False
38     False
39     False
40     False
41     False
43     False
45     False
46     False
48     False
52     False
53      True
54     False
55     False
56     False
57     False
58     False
59     False
60     False
61     False
62     False
64     False
67     False
70     False
71     False
72     False
73     False
74     False
75     False
76     False
77     False
79     False
81     False
83     False
84     False
85     False
86     False
87     False
89     False
90     False
91     False
93     False
94     False
97     False
99     False
100     True
101    False
102    False

**Removing Duplicates**

To remove duplicates, use the drop_duplicates() method.

In [ ]:
# Remove all duplicates:

df.drop_duplicates(inplace = True)

df

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

In [47]:
# Remove all duplicates (forestfires.csv):

# Code Here
df1.drop_duplicates(inplace = True)

df1

X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain  \
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0   
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0   
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0   
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2   
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0   
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0   
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0   
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0   
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0   
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0   
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0   
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3   38.0   4.0   0.0   
13   6  5   sep  mon  90.9  126.5  686.5   7.0  21.3   42.0   2.2   0.0   
14   6  5   sep  wed  92.9  133.3  699.6   9.2  26.4   21.0   4.5   0.0   
15   6  5   sep  fri  93.3  141.2  713.9  13.9  22.9   44.0   5.4   0.0   
17   8  5   oct  mon  84.9   32.8  664.2   3.0  16.7   47.0   4.9   0.0   
18   6  4   mar  wed  89.2   27.9   70.8   6.3  15.9   35.0   4.0   0.0   
20   6  4   sep  tue  91.0  129.5  692.6   7.0  18.3   40.0   2.7   0.0   
21   5  4   sep  mon  91.8   78.5  724.3   9.2  19.1   38.0   2.7   0.0   
22   7  4   jun  sun  94.3   96.3  200.0  56.1  21.0   44.0   4.5   0.0   
24   7  4   aug  sat  93.5  139.4  594.2  20.3  23.7   32.0   5.8   0.0   
25   7  4   aug  sun  91.4  142.4  601.4  10.6  16.3   60.0   5.4   0.0   
26   7  4   sep  fri  92.4  117.9  668.0  12.2  19.0   34.0   5.8   0.0   
28   6  3   sep  sat  93.4  145.4  721.4   8.1  30.2   24.0   2.7   0.0   
29   6  3   sep  sun  93.5  149.3  728.6   8.1  22.8   39.0   3.6   0.0   
31   6  3   sep  mon  88.6   91.8  709.9   7.1  11.2   78.0   7.6   0.0   
32   6  3   sep  fri  88.6   69.7  706.8   5.8  20.6   37.0   1.8   0.0   
33   6  3   sep  sun  91.7   75.6  718.3   7.8  17.7   39.0   3.6   0.0   
35   6  3   sep  tue  90.3   80.7  730.2   6.3  18.2   62.0   4.5   0.0   
36   6  3   oct  tue  90.6   35.4  669.1   6.7  21.7   24.0   4.5   0.0   
37   7  4   oct  fri  90.0   41.5  682.6   8.7  11.3   60.0   5.4   0.0   
38   7  3   oct  sat  90.6   43.7  686.9   6.7  17.8   27.0   4.0   0.0   
39   4  4   mar  tue  88.1   25.7   67.6   3.8  14.1   43.0   2.7   0.0   
40   4  4   jul  tue  79.5   60.6  366.7   1.5  23.3   37.0   3.1   0.0   
41   4  4   aug  sat  90.2   96.9  624.2   8.9  18.4   42.0   6.7   0.0   
43   4  4   sep  sat  92.5   88.0  698.6   7.1  19.6   48.0   2.7   0.0   
45   5  6   sep  wed  94.3   85.1  692.3  15.9  25.9   24.0   4.0   0.0   
46   5  6   sep  mon  90.9  126.5  686.5   7.0  14.7   70.0   3.6   0.0   
48   4  4   mar  mon  87.2   23.9   64.7   4.1  11.8   35.0   1.8   0.0   
52   4  3   aug  wed  92.1  111.2  654.1   9.6  20.4   42.0   4.9   0.0   
54   4  3   aug  thu  91.7  114.3  661.3   6.3  17.6   45.0   3.6   0.0   
55   4  3   sep  thu  92.9  137.0  706.4   9.2  27.7   24.0   2.2   0.0   
56   4  3   sep  tue  90.3   80.7  730.2   6.3  17.8   63.0   4.9   0.0   
57   4  3   oct  sun  92.6   46.5  691.8   8.8  13.8   50.0   2.7   0.0   
58   2  2   feb  mon  84.0    9.3   34.0   2.1  13.9   40.0   5.4   0.0   
59   2  2   feb  fri  86.6   13.2   43.0   5.3  12.3   51.0   0.9   0.0   
60   2  2   mar  sun  89.3   51.3  102.2   9.6  11.5   39.0   5.8   0.0   
61   2  2   mar  sun  89.3   51.3  102.2   9.6   5.5   59.0   6.3   0.0   
62   2  2   aug  thu  93.0   75.3  466.6   7.7  18.8   35.0   4.9   0.0   
64   2  2   aug  mon  91.1  103.2  638.8   5.8  23.1   31.0   3.1   0.0   
67   2  2   sep  fri  92.4  117.9  668.0  12.2  19.6   33.0   5.4   0.0   
70   4  5   mar  fri  91.2   48.3   97.8  12.5  15.8   27.0   7.6   0.0   
71   4  5   sep  fri  94.3   8

**Pandas - Data Correlations**

Finding Relationships
A great aspect of the Pandas module is the corr() method.

The corr() method calculates the relationship between each column in your data set.

The examples uses a CSV file called: 'data.csv'.

In [ ]:
# Show the relationship between the columns:

import pandas as pd

df = pd.read_csv('data.csv')

df.corr()

Duration     Pulse  Maxpulse  Calories
Duration  1.000000 -0.155408  0.009403  0.922717
Pulse    -0.155408  1.000000  0.786535  0.025121
Maxpulse  0.009403  0.786535  1.000000  0.203813
Calories  0.922717  0.025121  0.203813  1.000000

Note: The corr() method ignores "not numeric" columns.

**Result Explained**

The Result of the corr() method is a table with a lot of numbers that represents how well the relationship is between two columns.

The number varies from -1 to 1.

1 means that there is a 1 to 1 relationship (a perfect correlation), and for this data set, each time a value went up in the first column, the other one went up as well.

0.9 is also a good relationship, and if you increase one value, the other will probably increase as well.

-0.9 would be just as good relationship as 0.9, but if you increase one value, the other will probably go down.

0.2 means NOT a good relationship, meaning that if one value goes up does not mean that the other will.

\

**What is a good correlation?**

 It depends on the use, but I think it is safe to say you have to have at least 0.6 (or -0.6) to call it a good correlation.

**Perfect Correlation:**

We can see that "Duration" and "Duration" got the number 1.000000, which makes sense, each column always has a perfect relationship with itself.

**Good Correlation:**

"Duration" and "Calories" got a 0.922721 correlation, which is a very good correlation, and we can predict that the longer you work out, the more calories you burn, and the other way around: if you burned a lot of calories, you probably had a long work out.

**Bad Correlation:**

"Duration" and "Maxpulse" got a 0.009403 correlation, which is a very bad correlation, meaning that we can not predict the max pulse by just looking at the duration of the work out, and vice versa.

In [49]:
# Show the relationship between the columns (forestfires.csv):

# Code Here
import pandas as pd

df1 = pd.read_csv('forestfires.csv')

df1.corr()

<ipython-input-49-c530ba1d1794>:8: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df1.corr()


X         Y      FFMC       DMC        DC       ISI      temp  \
X     1.000000  0.539548 -0.021853 -0.052143 -0.079221  0.006917 -0.049092   
Y     0.539548  1.000000 -0.047157  0.004463 -0.091077 -0.023104 -0.025942   
FFMC -0.021853 -0.047157  1.000000  0.391315  0.346009  0.532487  0.428123   
DMC  -0.052143  0.004463  0.391315  1.000000  0.690180  0.314524  0.466866   
DC   -0.079221 -0.091077  0.346009  0.690180  1.000000  0.241454  0.504886   
ISI   0.006917 -0.023104  0.532487  0.314524  0.241454  1.000000  0.393684   
temp -0.049092 -0.025942  0.428123  0.466866  0.504886  0.393684  1.000000   
RH    0.090472  0.066762 -0.300783  0.062198 -0.048486 -0.135412 -0.538842   
wind  0.022777 -0.016028 -0.029907 -0.112474 -0.197004  0.106445 -0.221698   
rain  0.065594  0.033585  0.056928  0.075706  0.036977  0.067335  0.070067   
area  0.063385  0.044873  0.039801  0.071602  0.052112  0.007040  0.098721   

            RH      wind      rain      area  
X     0.090472  0.022777  0.065594  0.063385  
Y     0.066762 -0.016028  0.033585  0.044873  
FFMC -0.300783 -0.029907  0.056928  0.039801  
DMC   0.062198 -0.112474  0.075706  0.071602  
DC   -0.048486 -0.197004  0.036977  0.052112  
ISI  -0.135412  0.106445  0.067335  0.007040  
temp -0.538842 -0.221698  0.070067  0.098721  
RH    1.000000  0.077386  0.100403 -0.076414  
wind  0.077386  1.000000  0.061502  0.013340  
rain  0.100403  0.061502  1.000000 -0.007411  
area -0.076414  0.013340 -0.007411  1.000000